pkg/wireguard/agent/agent.go

```diff
@@ -732,97 +740,85 @@ type peerConfig struct {
-		delete(a.restoredAllowedIPs, pfx)
+// queuedAllowedIPUpdates returns the set of allowed IP insertions and removals
+// that are currently pending. If enableAllowedIPRemovals has not yet been
+// called, this method will not return any removals.
+func (p *peerConfig) queuedAllowedIPUpdates() (insert []net.IPNet, remove []net.IPNet) {
+	for _, ip := range p.needsInsert {
+		insert = append(insert, ip)
+	}
 
-		if !ok || prevKey == pubKey {
-			continue
-		}
+	for _, ip := range p.needsRemove {
+		remove = append(remove, ip)
+	}
 
-		// The restored allowed IP appears to be associated with a different
-		// peer. Hence, let's drop it from the other peer, to prevent flipping
-		// depending on which configuration gets applied last.
-		prevNode, ok := a.nodeNameByPubKey[prevKey]
-		if !ok {
-			continue
-		}
+	return
+}
 
-		prevPeer, ok := a.peerByNodeName[prevNode]
-		if !ok {
-			continue
-		}
+// hasAllowedIP returns true if ip has been synced to this peer on the device.
+func (p *peerConfig) hasAllowedIP(ip net.IPNet) bool {
+	_, exists := p.allowedIPs[ipnetToPrefix(ip)]
 
-		// It is not necessary that we push down this update, as WireGuard will
-		// automatically remove it from the old peer once it gets associated
-		// with the new one. However, we need to remove it from the local cache,
-		// to avoid reverting the change again upon a subsequent update.
-		prevPeer.removeAllowedIP(ipn)
-	}
+	return exists
 }
 
-// computeRestoredIPs computes and returns the set of restored but not yet observed
-// allowed IPs associated with the given peer. The computation is performed lazily
-// here, rather than directly as part of Agent.markRestoredIPsSeen, for performance
-// reasons, to limit the number of times we need to iterate over the list of restored
-// IPs and convert net.IPNet objects to netip.Prefix. Indeed, in the common case,
-// computeRestoredIPs is expected to be called twice for each peer, once during the
-// initial update (in which the majority of IPs is expected to be restored), and once
-// when restoration completes (when all IPs except stale ones need to be filtered).
-func (a *Agent) computeRestoredIPs(pubKey wgtypes.Key) []net.IPNet {
-	restored := a.restoredPeers[pubKey]
-	if len(restored) == 0 {
-		return nil
-	}
-
-	restored = slices.DeleteFunc(restored, func(r net.IPNet) bool {
-		_, found := a.restoredAllowedIPs[ipnetToPrefix(r)]
-		return !found
-	})
+// finishAllowedIPSync signals that any queued updates for the given ips have
+// been processed and synced to the device. This removes these ips from the
+// update queues.
+func (p *peerConfig) finishAllowedIPSync(ips []net.IPNet) {
+	for _, ip := range ips {
+		pfx := ipnetToPrefix(ip)
+		if aip, exists := p.needsInsert[pfx]; exists {
+			p.allowedIPs[pfx] = aip
+			delete(p.needsInsert, pfx)
+		}
 
-	a.restoredPeers[pubKey] = restored
-	return restored
+		if _, exists := p.needsRemove[pfx]; exists {
+			delete(p.allowedIPs, pfx)
+			delete(p.needsRemove, pfx)
+		}
+	}
```

Comment:
I wonder if it is worth setting the maps to `nil` if they are empty after we finished the sync. It will increase the number allocation/deallocations needed, but since otherwise the maps are never shrunk and basically consume memory forever.

Reply:
It probably wouldn't hurt. `UpdatePeer()` in particular has this part

```golang
peer.queueAllowedIPsInsert(a.ipCache.LookupByHostRLocked(nodeIPv4, nodeIPv6)...)
```

which would grow the capacity `needsInsert` well past what is needed later on. I do wonder about the effects of constant reallocation during periods of high pod churn.

Reply:
> I do wonder about the effects of constant reallocation during periods of high pod churn.

Yeah, it is a tradeoff. Ideally we'd cap the capacity of the map to a reasonable limit and reuse it as long as it's under the limit, and reset if if it's a above the limit.
Unfortunately it seems Golang has no way to inspect a map's capacity. We could track manually in a separate field which isn't pretty either. 

Not sure what to recommend here. I am worried that in particular with large-scale clustermesh setups, we might have up to 100k-300k of queued allowed IPs in total, meaning leaking about 10-30 MBs of memory.

(unrelated: we've also discussed adding pod CIDRs (where possihble) rather than individal pod IPs as allowedIPs, that would also reduce the pod churn issue. But that's not something we should tackle in this PR)

Reply:
> Unfortunately it seems Golang has no way to inspect a map's capacity.

You could potentially just use the map's length as a proxy for its capacity inside `finishAllowedIPSync()` to determine whether or not the capacity has reached a "high water mark". I think we should probably just start simple with setting the maps to `nil` every time here though. I think in the average case memory consumption is going to be the main thing we are concerned with.

Reply:
> unrelated: we've also discussed adding pod CIDRs (where possihble) rather than individal pod IPs as allowedIPs,

I was thinking about this as well. Are there any IPAM modes where there isn't a clear CIDR range per node where this approach wouldn't work? I suspect you may even get some performance benefit with WireGuard throughput this way, as it means quicker lookups in the driver's internal routing table to determine the peer associated with an IP. It would be interesting to test this out.

Reply:
> I was thinking about this as well. Are there any IPAM modes where there isn't a clear CIDR range per node where this approach wouldn't work? 

Yes, there are. ENI mode[^1], delegated IPAM mode, manual CRD mode, legacy Azure IPAM and AlibabaCloud all do not have the concept of a pod CIDR. Instead, IPs are allocated from a global CIDR that is shared among all nodes. On such systems, we will have to keep per-IP lists. 

But for IPAM modes `cluster-pool` (default), `kubernetes` (common, default on GKE) and `multi-pool`[^2] we could absolutely switch to a podCIDR based approach.

> It would be interesting to test this out.

That would be awesome. I'm more than happy to review such a PR.

[^1]: Some ENI nodes do run with prefix delegation, which means pod IPs are allocated out of /28 CIDRs on each node (with multiple CIDRs per node). That would still reduce the number of entries in "allowedIPs". However, those /28 CIDRs are not exposed to the agent's node manager at the moment.
[^2]: `multi-pool` will have multiple CIDRs per node that are allocated dynamically, so it's a bit more involved than `cluster-pool` or `kubernetes` which have a single static pod CIDR. But still pretty much doable.

Reply:
> I think we should probably just start simple with setting the maps to nil every time here though.

Done in the latest push.